lib/domain.js

```diff
@@ -399,32 +399,35 @@ EventEmitter.init = function() {
 const eventEmit = EventEmitter.prototype.emit;
 EventEmitter.prototype.emit = function emit(...args) {
   const domain = this.domain;
-  if (domain === null || domain === undefined || this === process) {
-    return Reflect.apply(eventEmit, this, args);
-  }
 
   const type = args[0];
-  // edge case: if there is a domain and an existing non error object is given,
-  // it should not be errorized
-  // see test/parallel/test-event-emitter-no-error-provided-to-error-event.js
-  if (type === 'error' && args.length > 1 && args[1] &&
-    !(args[1] instanceof Error)) {
-    domain.emit('error', args[1]);
-    return false;
-  }
+  const shouldEmitError = type === 'error' &&
+                          this.listenerCount(type) > 0;
 
-  domain.enter();
-  try {
+  // Just call original `emit` if current EE instance has `error`
+  // handler, there's no active domain or this is process
+  if (shouldEmitError || domain === null || domain === undefined ||
+      this === process) {
     return Reflect.apply(eventEmit, this, args);
-  } catch (er) {
+  }
+
+  if (type === 'error') {
+    const er = args.length > 1 && args[1] ?
+      args[1] : new errors.Error('ERR_UNHANDLED_ERROR');
+
     if (typeof er === 'object' && er !== null) {
```

Comment:
@BridgeAR Great job noticing that! Seems like we've had that pointless check in this code for a while. Guessing it was useful at some point but clearly left over and unnecessary now.

The only reason I can think of leaving the object check is in case someone is testing for those properties and assuming that they're dealing with an object afterwards... ? But that seems obscure enough that it shouldn't be a real issue.

Reply:
That could be solved with a comment as well, right?

Reply:
I just mean for the purposes of third-party code... but that seems like a pretty obscure thing to me. We can run CitGM to double-check.

Reply:
Although, something to think about: `typeof` isn't very expensive in V8 these days and setting 3 props on numbers or strings is, I think, a lot more expensive.

Reply:
I am not sure what you mean that third-party code relies on that? Trying to set a property on a primitive (not function) is a noop.

About the performance implications - I am aware that a typeof is cheap but do we want to optimize for the average case or for obscure cases? Because passing in a non object as error sounds really weird to me...

Reply:
Does it? I've definitely seen people use strings. It's tricky... we've allowed it till now so we're kind of stuck with supporting it. Anyway, if we're optimizing then it should be done in another PR. I don't think `domain` related code is that performance sensitive anyway... there are much worse offenders.